### ISSUE TITLE
generic is incorrectly inferred as invariant if class uses `Self` type on an attribute

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#18334]
```py
from typing import Self


class Foo[T]:
    instance: Self # commenting out this line makes the error go away

    def foo(self) -> T: ... # type:ignore[empty-body]


class Bar(Foo[int]): ...

# Incompatible types in assignment (expression has type "Bar", variable has type "Foo[object]")  [assignment]
foo: Foo[object] = Bar()
```
[playground](https://mypy-play.net/?mypy=latest&python=3.13&flags=show-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=efefed751bdb0d6af3282c0ba018a5f8)

**Comment by ilevkivskyi** [comment]
```python
class Foo[T]:
    instance: Self
```
is invariant for the same reason
```python
class Foo[T]:
    instance: Foo[T]
```
is invariant: `instsance` is a settable attribute. So this is actually true positive.

**Comment by DetachHead** [comment]
hmm interesting. here's a simplified example that mypy doesn't catch:

```py
from typing import Self, reveal_type

class Foo:
    instance: Self

class Bar(Foo): ...

bar = Bar()

def fn(value: Foo):
    foo: Foo = value
    foo.instance = Foo()

fn(bar)

reveal_type(bar.instance) # Bar, but Foo at runtime
```

[playground](https://mypy-play.net/?mypy=latest&python=3.13&flags=show-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=95f3803db7523bf6189880ca0f720fa3)

**Comment by ilevkivskyi** [comment]
This is a well known unsafety of self-types. In fact, essentially same problem existed long before `Self` notation was even introduced, for example this is unsafe:
```python
from typing import TypeVar

T = TypeVar("T")
class Base:
    def foo(self: T, arg: T) -> None: ...

class Sub(Base):
    x: int
    def foo(self, arg: Sub) -> None:
        arg.x

b: Base = Sub()
b.foo(Base())
```
Most (if not all) false negatives with self types boil down to this hole in type system. However, the current consensus is that it is impossible to fix it without making self-types a pain to work with.

**Comment by DetachHead** [comment]
my proposed solution is to add new rules to:

1. ban `Self` from being used in a contravariant position (except for the `self` parameter on instance methods) (https://github.com/DetachHead/basedpyright/issues/1586)
2. ban calling instance methods on the class (https://github.com/DetachHead/basedpyright/issues/1566)

i've found that there is no downside to adding new diagnostic rules for stuff like this, because users who want the additional safety can benefit from them, but users who think it would make self-types a pain to work with can just keep those rules disabled.

**Comment by ilevkivskyi** [comment]
Yeah, adding a new flag (or if possible an error code) may be a solution. This still however requires some discussions on exact semantics. Also to give correct expectations: since this hole causes (surprisingly) few problems in real-world code, fixing this is not high priority.

If you are interested in this, I guess https://github.com/python/mypy/issues/15619 may be a better place to continue discussion.

**Comment by KotlinIsland** [comment]
```py
from typing import TypeVar

class Base:
    def foo(self) -> None: ...

class Sub(Base):
    x: int = 1
    def foo(self) -> None:
        self.x

b: type[Base] = Sub
b.foo(Base())
```

**Comment by KotlinIsland** [comment]
> class Foo[T]:
>     instance: Self
> is invariant for the same reason
> 
> class Foo[T]:
>     instance: Foo[T]
> is invariant: `instsance` is a settable attribute. So this is actually true positive.


why is there no error here then?

```py
from typing import Self, Generic, TypeVar

out_T = TypeVar("out_T", covariant=True)

class Foo(Generic[out_T]):
    instance: Self

    def f(self) -> out_T: ...
    
f1 = Foo[int]()
f2: Foo[object] = f1
```

what exactly does this help with? making it invariant because of `Self` usage doesn't prevent any invalid usages

**Comment by jorenham** [comment]
From the [typing spec](https://typing.python.org/en/latest/spec/generics.html#self):

```py
from typing import Self

@dataclass
class LinkedList(Generic[T]):
    value: T
    next: Self | None = None

@dataclass
class OrdinalLinkedList(LinkedList[int]):
    def ordinal_value(self) -> str:
        return as_ordinal(self.value)

xs = OrdinalLinkedList(value=1, next=LinkedList[int](value=2))
# Type error: Expected OrdinalLinkedList, got LinkedList[int].

if xs.next is not None:
    xs.next = OrdinalLinkedList(value=3, next=None)  # OK
    xs.next = LinkedList[int](value=3, next=None)  # Not OK
```

The code above is semantically equivalent to treating each attribute containing a `Self` type as a `property` that returns that type:

```py
from dataclasses import dataclass
from typing import Any, Generic, TypeVar

T = TypeVar("T")
Self = TypeVar("Self", bound="LinkedList")


class LinkedList(Generic[T]):
    value: T

    @property
    def next(self: Self) -> Self | None:
        return self._next

    @next.setter
    def next(self: Self, next: Self | None) -> None:
        self._next = next

class OrdinalLinkedList(LinkedList[int]):
    def ordinal_value(self) -> str:
        return str(self.value)
```

---

So the example in the OP is *specified* to be equivalent to:

```py
class Foo[T]:
    @property
    def instance[SelfT: Foo[Any]](self) -> Self: ...
    @instance.setter
    def instance[SelfT: Foo[Any]](self, instance: SelfT, /) -> None: ...

    def foo(self) -> T: ... # type:ignore[empty-body]

class Bar(Foo[int]):
    @property
    def instance[SelfT: Bar[Any]](self) -> Self: ...
    @instance.setter
    def instance[SelfT: Bar[Any]](self, instance: SelfT, /) -> None: ...
```

Mypy seems to assume that `Self` is class-scoped. It isn't; `Self` is method- or property-scoped. So `Self` cannot influence the (variance of) class-scoped type parameters.

**Comment by randolf-scholz** [comment]
> Mypy seems to assume that `Self` is class-scoped. It isn't; `Self` is method- or property-scoped. So `Self` cannot influence the (variance of) class-scoped type parameters.

Wouldn't this mean that `ClassVar[Self]` is illegal? This is actually part of the [test-suite](https://github.com/python/mypy/blob/0ece662da24111c25345fdba5e69346bbd5c287f/test-data/unit/check-selftype.test#L1655-L1660). Your comment also gave me the idea to try the following, since `mypy` sometimes treat an annotation inside `__init__` differently than an annotation inside the class-body.[^1] This led to an odd result [playground](https://mypy-play.net/?mypy=master&python=3.12&gist=4f753938f810b5d2bf69ec78931fba0d)

```python
from typing import Self

class Foo:
    def __init__(self) -> None:
        self.instance: Self = self
# main.py:5: error: Incompatible types in assignment (expression has type "Foo", variable has type "Self")  [assignment]
```




[^1]: as an annotated attribute inside the class body can be a class- or instance-attribute whereas a `self.attr` annotated inside `__init__` is always treated as an instance-only attribute. Though I sometimes wish that were the default for everything that's not explicitly marked as `ClassVar`.

**Comment by jorenham** [comment]
> Wouldn't this mean that `ClassVar[Self]` is illegal?

Although not explicitly stated in the typing spec, it would be legal if you think of class-attributes as a "classproperty", i.e. a `@property` + `@classmethod`, analogous to instance attributes -> `@property`
